### DISCUSSION TITLE
Change Input/Output Device from command line/script

### DISCUSSION THREAD

**Discussion by Ameb** [wwmm/easyeffects#1337] [Q&A]
Hello.
I use multiple audio devices. keep all of them connected and use the Input / Output device dropdown to switch between them (Easyeffects is my default device in pipewire)

```
$ pactl get-default-sink   
easyeffects_sink
```

Is it possible to change the input/output device from CLI or a script? We can bind presets to devices, but not the other way around.

A screenshot to understand what i mean with the device dropdown:

![image](https://user-images.githubusercontent.com/386567/149456326-bf56f55a-ae9e-4bd8-a20e-a9b725048b6e.png)

Cheers

**Comment by Ameb** [comment]
As a workaround i was able to map the output of easyeffects (one of the `output_level` pipewire nodes) to the desired device by using `pw-link`.

The hardest part was locating the proper output from easyeffects, as it creates many nodes for its filters.
![image](https://user-images.githubusercontent.com/386567/149643520-d026dc9c-b6ae-41ef-9e78-e6ed81c39e4b.png)


Having a patchbay like `Helvum` (catia, etc should also work) can help with locating the proper input and output nodes/ports. I used Helvum to reroute the nodes as desired with `pw-cli` open in a terminal to watch the connection events

```
remote 0 removed global: 	id 165, type PipeWire:Interface:Link/3
 		object.serial = "7640"
 		factory.id = "20"
 		link.output.port = "91"
 		link.input.port = "72"
 		link.output.node = "88"
 		link.input.node = "53"
```
The `port` is the interesting number. Then use `pw-cli` to list the ports and locate yours
```
$ pw-link -iI | grep 72                                                                                                                                                                                                                                                          
  72 alsa_output.pci-0000_0c_00.1.hdmi-stereo-extra1:playback_FL
$ pw-link -oI | grep 91                                                                                                                                                                                                                                                          
  91 ee_soe_output_level:output_FL
```
Automate the routing as needed with `pw-link` commands.
```
pw-link ee_soe_output_level:output_FL alsa_output.pci-0000_0c_00.1.hdmi-stereo-extra1:playback_FL
pw-link ee_soe_output_level:output_FR alsa_output.pci-0000_0c_00.1.hdmi-stereo-extra1:playback_FR
```

**Reply by wwmm** [reply]
If you want to write a script follow what I suggest at https://github.com/wwmm/easyeffects/discussions/916. You can find out the key paths using dconf-editor
![Screenshot from 2022-01-16 01-47-06](https://user-images.githubusercontent.com/16225172/149647641-f8f17ce5-31de-45e5-b95b-87786ac6b349.png)

**Reply by Ameb** [reply]
I wish i could mark your comment as answer... thank you!

**Comment by srosato** [comment]
For those who installed easyeffects as a flatpak the way I did it is like this

`pactl list short sinks`

In my case I had 

```
55      alsa_output.usb-Generic_USB_Audio-00.analog-stereo      PipeWire        s32le 2ch 48000Hz       IDLE
74      easyeffects_sink        PipeWire        float32le 2ch 48000Hz   RUNNING
1611351 bluez_output.00_1B_66_F9_4D_80.a2dp-sink        PipeWire        s24le 2ch 48000Hz       RUNNING
```
Where `alsa_output.usb-Generic` was my speakers and `bluez_output` was my bluetooth headset

I then ran the command `flatpak run --command=gsettings com.github.wwmm.easyeffects set com.github.wwmm.easyeffects.streamoutputs output-device bluez_output.00_1B_66_F9_4D_80.a2dp-sink` in order for easyeffects to switch to my bluetooth headset. I created a keybdinding using my window manager (i3 in my case)

**Comment by Rdkang** [comment]
I have adapted this script and it works great https://github.com/onezerodan/PipeWire-switch-output.

hope this helps :)
